File: backend/app/api/v1/endpoints/tasks.py
from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session

from app.api.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.task import (
    TaskCreate,
    TaskUpdate,
    TaskResponse,
)
from app.services.task_service import (
    create_task,
    get_tasks,
    get_task,
    update_task,
    delete_task,
)

router = APIRouter(
    prefix="/tasks",
    tags=["Tasks"],
)


@router.post(
    "",
    response_model=TaskResponse,
    status_code=status.HTTP_201_CREATED,
)
def create_new_task(
    task: TaskCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    created_task = create_task(db, current_user, task)

    if created_task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Project not found",
        )

    return created_task


@router.get(
    "",
    response_model=list[TaskResponse],
)
def list_tasks(
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return get_tasks(db, current_user)


@router.get(
    "/{task_id}",
    response_model=TaskResponse,
)
def get_single_task(
    task_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = get_task(db, current_user, task_id)

    if task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )

    return task


@router.put(
    "/{task_id}",
    response_model=TaskResponse,
)
def edit_task(
    task_id: UUID,
    task_data: TaskUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = get_task(db, current_user, task_id)

    if task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )

    return update_task(db, task, task_data)


@router.delete(
    "/{task_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
def remove_task(
    task_id: UUID,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    task = get_task(db, current_user, task_id)

    if task is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail="Task not found",
        )

    delete_task(db, task)

File: frontend/src/api/tasks.ts
import api from "./axios";

export interface TaskCreate {
  project_id: string;
  title: string;
  description?: string;
  status?: "Todo" | "In Progress" | "Completed";
  priority?: "Low" | "Medium" | "High";
  due_date?: string | null;
}

export interface TaskUpdate {
  title?: string;
  description?: string;
  status?: "Todo" | "In Progress" | "Completed";
  priority?: "Low" | "Medium" | "High";
  due_date?: string | null;
}

export async function getTasks() {
  const response = await api.get("/tasks");
  return response.data;
}

export async function createTask(
  task: TaskCreate
) {
  const response = await api.post(
    "/tasks",
    task
  );

  return response.data;
}

export async function updateTask(
  id: string,
  task: TaskUpdate
) {
  const response = await api.put(
    `/tasks/${id}`,
    task
  );

  return response.data;
}

export async function deleteTask(id: string) {
  await api.delete(`/tasks/${id}`);
}